File: apps/mobile-flutter/test/shared/widgets/gauge/gauge_arc_geometry_test.dart
import 'dart:math' as math;
import 'dart:ui';

import 'package:flutter_test/flutter_test.dart';
import 'package:kallo_mobile/shared/widgets/gauge/gauge_arc_geometry.dart';

/// The reference dial, read off the rendered source component rather than
/// re-derived here: centre (268, 96), inner 54, outer 72, cornerRadius 8,
/// paddingAngle 4, 210° → −30°, filled to 84%.
///
/// Its filled segment begins at 213.0091,127.749 — the point where the first
/// corner circle meets the sector's radial edge. Every expectation below is one
/// of that path's own numbers, so this test fails if our geometry drifts from
/// the shape it was ported from.
const _center = Offset(268, 96);
const _outerRadius = 72.0;

Offset _startOf(Path path) =>
    path.computeMetrics().first.getTangentForOffset(0)!.position;

/// Walk the outline and take the real extremes.
///
/// `Path.getBounds()` is no use here: for arcs it returns the bounds of the
/// CONTROL points, which sit well outside a curve this round — it reported a
/// left edge 23px beyond the dial's own radius.
Rect _tightBounds(Path path) {
  var left = double.infinity, top = double.infinity;
  var right = double.negativeInfinity, bottom = double.negativeInfinity;
  for (final metric in path.computeMetrics()) {
    for (var i = 0; i <= 400; i++) {
      final p = metric.getTangentForOffset(metric.length * i / 400)!.position;
      left = math.min(left, p.dx);
      top = math.min(top, p.dy);
      right = math.max(right, p.dx);
      bottom = math.max(bottom, p.dy);
    }
  }
  return Rect.fromLTRB(left, top, right, bottom);
}

void main() {
  group('gaugePaths', () {
    test('starts where the source component starts', () {
      final paths = gaugePaths(
        center: _center,
        outerRadius: _outerRadius,
        progress: 0.84,
      );
      final start = _startOf(paths.filled);
      expect(start.dx, closeTo(213.0091, 0.05));
      expect(start.dy, closeTo(127.749, 0.05));
    });

    test('spans the reference sector', () {
      final bounds = _tightBounds(gaugePaths(
        center: _center,
        outerRadius: _outerRadius,
        progress: 0.84,
      ).filled);
      // Left and top are the sector's own extremes: 180° and 90° at radius 72.
      expect(bounds.left, closeTo(_center.dx - _outerRadius, 0.05));
      expect(bounds.top, closeTo(_center.dy - _outerRadius, 0.05));
      // The far end stops short of the full circle — 84%, not 100%.
      expect(bounds.right, lessThan(_center.dx + _outerRadius));
      expect(bounds.right, closeTo(336.1, 0.6));
    });

    test('the remainder is a separate pill, not a track behind the fill', () {
      final paths = gaugePaths(
        center: _center,
        outerRadius: _outerRadius,
        progress: 0.84,
      );
      // A 4° gap sits between them, so the two never share an edge.
      expect(paths.filled.getBounds().overlaps(paths.remainder.getBounds()),
          isTrue,
          reason: 'bounding boxes still overlap — they are arcs, not blocks');
      final gap = _startOf(paths.remainder) - _startOf(paths.filled);
      expect(gap.distance, greaterThan(1));
    });

    test('drops the segment that has no sweep left', () {
      final empty = gaugePaths(
        center: _center,
        outerRadius: _outerRadius,
        progress: 0,
      );
      expect(empty.filled.computeMetrics().isEmpty, isTrue);
      expect(empty.remainder.computeMetrics().isEmpty, isFalse);

      final full = gaugePaths(
        center: _center,
        outerRadius: _outerRadius,
        progress: 1,
      );
      expect(full.filled.computeMetrics().isEmpty, isFalse);
      expect(full.remainder.computeMetrics().isEmpty, isTrue);
    });

    test('clamps past target rather than winding past the sweep', () {
      final over = gaugePaths(
        center: _center,
        outerRadius: _outerRadius,
        progress: 1.6,
      );
      final full = gaugePaths(
        center: _center,
        outerRadius: _outerRadius,
        progress: 1,
      );
      expect(over.filled.getBounds(), full.filled.getBounds());
    });

    test('holds its proportions when scaled', () {
      // The macro dials are the same shape at 44. Band and corner are ratios,
      // so the small dial's bounds are the reference's, scaled.
      final small = gaugePaths(
        center: Offset.zero,
        outerRadius: 36,
        progress: 0.84,
      ).filled.getBounds();
      final large = gaugePaths(
        center: Offset.zero,
        outerRadius: 72,
        progress: 0.84,
      ).filled.getBounds();
      expect(small.width * 2, closeTo(large.width, 0.05));
      expect(small.height * 2, closeTo(large.height, 0.05));
    });
  });

  test('gaugeTipOffset puts the tips half a radius below centre', () {
    // 210° and −30° both sit at sin = −0.5, so the tips are at r/2 below the
    // centre. The readout leans on this to sit level with them.
    expect(gaugeTipOffset(104), 52);
    final tip = arcPoint(_center, _outerRadius, kGaugeEndAngle);
    expect(tip.dy - _center.dy, closeTo(gaugeTipOffset(_outerRadius), 0.001));
  });
}

File: lib/core/ui/__tests__/gauge-arc-geometry.test.ts
import { describe, expect, it } from 'vitest';
import {
  arcPoint,
  GAUGE_END_ANGLE,
  gaugeHeight,
  gaugePaths,
  gaugeTipOffset,
} from '@/lib/core/ui/gauge-arc-geometry';

/**
 * The reference dial, read off the rendered source component rather than
 * re-derived here: centre (268, 96), inner 54, outer 72, cornerRadius 8,
 * paddingAngle 4, 210° → −30°, filled to 84%.
 *
 * These are the SAME numbers `gauge_arc_geometry_test.dart` pins the Flutter
 * implementation to. The two platforms draw this dial from two files, and the
 * only thing keeping them the same shape is that both answer to this fixture —
 * so if you change one, change both, and if you cannot, the drift is real.
 */
const CENTER = { x: 268, y: 96 };
const OUTER_RADIUS = 72;

interface Segment {
  op: 'M' | 'L' | 'A';
  /** Arc radius; absent on move/line. */
  radius?: number;
  /** The two arc flags, which are shape, not size — they never scale. */
  flags?: [number, number];
  x: number;
  y: number;
}

/**
 * The path as its commands, so a claim can be made about the points without
 * the `A` command's rotation and flags being mistaken for coordinates.
 */
function parsePath(path: string): Segment[] {
  return path
    .split(/(?=[MLAZ])/)
    .map((part) => part.trim())
    .filter((part) => part !== '' && part !== 'Z')
    .map((part) => {
      const [op, ...rest] = part.split(/\s+/);
      const n = rest.map(Number);
      if (op === 'A') {
        const [radius, , , largeArc, sweep, x, y] = n;
        return {
          op: 'A' as const,
          radius,
          flags: [largeArc, sweep] as [number, number],
          x,
          y,
        };
      }
      return { op: op as 'M' | 'L', x: n[0], y: n[1] };
    });
}

/** The `M x y` the path opens on — the point the first corner circle meets. */
function startOf(path: string): { x: number; y: number } {
  const [first] = parsePath(path);
  return { x: first.x, y: first.y };
}

/**
 * How wide the path's own points span. Only ever compared against another
 * width from the same fixture, so the arcs bulging past their endpoints is a
 * constant, not an error.
 */
function spanOf(path: string): number {
  const xs = parsePath(path).map((segment) => segment.x);
  return Math.max(...xs) - Math.min(...xs);
}

describe('gaugePaths', () => {
  it('starts where the source component starts', () => {
    const { filled } = gaugePaths(CENTER, OUTER_RADIUS, 0.84);
    const start = startOf(filled);

    expect(start.x).toBeCloseTo(213.0091, 2);
    expect(start.y).toBeCloseTo(127.749, 2);
  });

  it('reaches the sector’s own extremes and stops short of the circle', () => {
    const { filled } = gaugePaths(CENTER, OUTER_RADIUS, 0.84);
    const xs = parsePath(filled).map((segment) => segment.x);

    // 180° at radius 72 is the sector's left edge; nothing may sit outside it.
    expect(Math.min(...xs)).toBeGreaterThanOrEqual(CENTER.x - OUTER_RADIUS);
    // The far end stops short of the full circle — 84%, not 100%.
    expect(Math.max(...xs)).toBeLessThan(CENTER.x + OUTER_RADIUS);
  });

  it('draws the remainder as a separate pill, not a track behind the fill', () => {
    const { filled, remainder } = gaugePaths(CENTER, OUTER_RADIUS, 0.84);
    const gap = Math.hypot(
      startOf(remainder).x - startOf(filled).x,
      startOf(remainder).y - startOf(filled).y
    );

    // A 4° gap sits between them, so the two never share an edge.
    expect(gap).toBeGreaterThan(1);
  });

  it('drops the segment that has no sweep left', () => {
    const empty = gaugePaths(CENTER, OUTER_RADIUS, 0);
    expect(empty.filled).toBe('');
    expect(empty.remainder).not.toBe('');

    const full = gaugePaths(CENTER, OUTER_RADIUS, 1);
    expect(full.filled).not.toBe('');
    expect(full.remainder).toBe('');
  });

  it('draws a fill for a barely-started day', () => {
    // 60 kcal of 1,844 — the day that reported this bug. Its sweep is 7.7°,
    // far under the ~15.3° the nominal corners need, so the fill used to be
    // dropped and every dial read as untouched.
    for (const radius of [104, 44, 30]) {
      expect(
        gaugePaths(CENTER, radius, 60 / 1844).filled,
        `no fill at 3.25% on a dial of radius ${radius}`
      ).not.toBe('');
    }
  });

  it('keeps small values proportional rather than snapping to a floor', () => {
    const spans = [0.03, 0.04, 0.06, 0.1].map((progress) =>
      spanOf(gaugePaths(CENTER, OUTER_RADIUS, progress).filled)
    );
    for (let i = 1; i < spans.length; i++) {
      expect(spans[i]).toBeGreaterThan(spans[i - 1]);
    }
  });

  it('leaves a nearly-finished day the track it has left', () => {
    // The mirror of the same guard: above ~93.5% the remainder used to be
    // dropped, so a 96% day rendered as a completely full dial.
    const paths = gaugePaths(CENTER, OUTER_RADIUS, 0.96);
    expect(paths.remainder).not.toBe('');
    expect(paths.filled).not.toBe('');
  });

  it('holds its proportions when scaled at a small value', () => {
    // The fitted corner is derived from the radii, so it has to stay a pure
    // ratio down where it is doing the fitting.
    const small = parsePath(gaugePaths({ x: 0, y: 0 }, 36, 0.03).filled);
    const large = parsePath(gaugePaths({ x: 0, y: 0 }, 72, 0.03).filled);
    expect(large.length).toBeGreaterThan(0);
    expect(small).toHaveLength(large.length);
    small.forEach((segment, i) => {
      expect(segment.x * 2).toBeCloseTo(large[i].x, 2);
      expect(segment.y * 2).toBeCloseTo(large[i].y, 2);
    });
  });

  it('draws nothing for a zero target rather than a path of NaNs', () => {
    const paths = gaugePaths(CENTER, OUTER_RADIUS, Number.NaN);
    expect(paths.filled).toBe('');
    expect(paths.remainder).not.toContain('NaN');
  });

  it('clamps past target rather than sweeping round twice', () => {
    expect(gaugePaths(CENTER, OUTER_RADIUS, 1.6).filled).toBe(
      gaugePaths(CENTER, OUTER_RADIUS, 1).filled
    );
  });

  it('holds its proportions when scaled', () => {
    const origin = { x: 0, y: 0 };
    const small = parsePath(gaugePaths(origin, 36, 0.84).filled);
    const large = parsePath(gaugePaths(origin, 72, 0.84).filled);

    expect(small).toHaveLength(large.length);
    for (const [i, segment] of small.entries()) {
      // Same commands, same flags, every length doubled — one shape at two
      // sizes, which is what lets the macro dials be the calorie dial small.
      expect(segment.op).toBe(large[i].op);
      expect(segment.flags).toEqual(large[i].flags);
      expect(segment.x * 2).toBeCloseTo(large[i].x, 2);
      expect(segment.y * 2).toBeCloseTo(large[i].y, 2);
      if (segment.radius !== undefined) {
        expect(segment.radius * 2).toBeCloseTo(large[i].radius as number, 2);
      }
    }
  });
});

describe('gaugeTipOffset', () => {
  it('is where the arc actually ends', () => {
    expect(gaugeTipOffset(104)).toBe(52);
    // 210° and −30° both sit at sin = −0.5, so the tips are exactly r/2 below
    // the centre — the offset is that fact, not a tuned constant.
    expect(
      arcPoint(CENTER, OUTER_RADIUS, GAUGE_END_ANGLE).y - CENTER.y
    ).toBeCloseTo(gaugeTipOffset(OUTER_RADIUS), 2);
  });
});

describe('gaugeHeight', () => {
  it('covers the cap that hangs below the tips', () => {
    expect(gaugeHeight(72)).toBeCloseTo(117, 5);
    expect(gaugeHeight(52)).toBeCloseTo(84.5, 5);
  });
});
